DatadogRUM/Sources/Instrumentation/AppHangs/AppHangsWatchdogThread.swift

```diff
@@ -0,0 +1,131 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-Present Datadog, Inc.
+ */
+
+import Foundation
+import DatadogInternal
+
+internal struct AppHang {
+    /// The date of hang end.
+    let date: Date
+    /// The duration of the hang.
+    let duration: TimeInterval
+    // TODO: RUM-2925 Add hang stack trace
+}
+
+internal final class AppHangsWatchdogThread: Thread {
+    enum Constants {
+        /// The "idle" interval for sleeping the watchdog thread before scheduling the next task on the main queue, represented as a percentage of the `appHangThreshold`.
+        ///
+        /// Introducing even a small sleep significantly reduces CPU consumption by the watchdog thread (nearly 0%). No sleep would result in close to 100% CPU usage.
+        ///
+        /// `tolerance` defines the margin of error in hang monitoring. The "measured hang duration" may differ from `0` to `tolerance` % when compared to the "actual hang duration".
+        /// This means some actual hangs very close to `appHangThreshold` might be considered false negatives and not reported.
+        ///
+        /// **Example:**
+        /// If `appHangThreshold` is 2 seconds and tolerance is 2.5%, the watchdog thread will idle for 50 milliseconds before taking each sample of the main thread.
+        /// Because the sleep might start right at the moment of hang beginning on the main thread, the watchdog thread will miss the first 50 milliseconds of hang duration.
+        /// If the actual hang lasts exactly 2 seconds, the watchdog will measure it as 1950 milliseconds. As a result, the SDK will not report it as it falls under `appHangThreshold`.
+        static let tolerance: Double = 0.025 // 2.5%
+    }
+
+    /// The minimal duration of the main thread hang to consider it an App Hang.
+    private let appHangThreshold: TimeInterval
+    /// The "idle" interval for watchdog thread operations. Reduces the pressure on CPU.
+    private let idleInterval: TimeInterval
+    /// Semaphore used to block this thread until main thread responds.
+    private let mainThreadTask = DispatchSemaphore(value: 0)
+    /// The queue to observe by this thread (main queue).
+    private let mainQueue: DispatchQueue
+    /// SDK date provider.
+    private let dateProvider: DateProvider
+    /// Telemetry interface.
+    private let telemetry: Telemetry
```

Comment:
curious, having strong reference to Telemetry, can it cause reference cycle and in the end memory leak here?

Reply:
This is implemented by [`CoreTelemetry`](https://github.com/DataDog/dd-sdk-ios/blob/ab930187170a1170d5131e0992f81f47ad5fd280/DatadogInternal/Sources/Telemetry/Telemetry.swift#L302-L304) which weakifies the underlying `core` reference.